keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,308 @@
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+(**FIXME:** Explain what is the current situation and then roughly explain what this feature does)
+
+> This section is incredibly important for producing high quality user-focused
+> documentation such as release notes or a development roadmap.  It should be
+> possible to collect this information before implementation begins in order to
+> avoid requiring implementors to split their attention between writing release
+> notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+> should help to ensure that the tone and content of the `Summary` section is
+> useful for a wide audience.
+>
+> A good summary is probably at least a paragraph in length.
+
+## Motivation
+
+(**FIXME:** Explain that we don't want to copy-paste stuff, list issues that wanted it at some form)
+
+> This section is for explicitly listing the motivation, goals and non-goals of
+> this KEP.  Describe why the change is important and the benefits to users.  The
+> motivation section can optionally provide links to [experience reports][] to
+> demonstrate the interest in a KEP within the wider Kubernetes community.
+>
+> [experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+(**FIXME:** Reusable configuration, anything else?)
+
+> List the specific goals of the KEP.  What is it trying to achieve?  How will we
+> know that this has succeeded?
+
+### Non-Goals
+
+(**FIXME:** What to add here?)
+
+> What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+> and make progress.
+
+## Proposal
+
+(**FIXME:** High-level explanation of how it works)
+
+> This is where we get down to the specifics of what the proposal actually is.
+> This should have enough detail that reviewers can understand exactly what
+> you're proposing, but should not include things like API designs or
+> implementation.  The "Design Details" section below is for the real
+> nitty-gritty.
+
+### User Story
+
+(**FIXME:** Write here the example from the PR, minus the section for the current situation)
+
+> Detail the things that people will be able to do if this KEP is implemented.
+> Include as much detail as possible so that people can understand the "how" of
+> the system.  The goal here is to make this feel real for users without getting
+> bogged down.
+
+
+### Notes/Constraints/Caveats (optional)
```

Comment:
I think we can remove this section altogether, unless someone wants to add something. I do see a constraint when generators are used with components, but the more I think of it, the more it looks orthogonal to this KEP.

Reply:
Come to think of it, we can add the following:

The distinction between components and overlays is evident when transformers are involved, but not when a component adds resources only. This may confuse users who will not know what to use in this case, and must be communicated clearly in the docs.

Reply:
What about custom transformers? I have never tested them and, as such, I am not sure whether there are implications when:

- one uses both `components` and a custom transformer in the same kustomization
- one use a `component` that uses a custom transformer

Reply:
I don't think that there should be an implication, since only the resource accumulation is affected, and the transformers will simply work on top of that. @pgpx, maybe you know something more on this?

Reply:
Yes, transformers should just work because they are only accumulated after resources (and now components) - see [this part of the code](https://github.com/kubernetes-sigs/kustomize/blob/c5c53011da134543c604d0d8047d22cce1a0aa12/api/internal/target/kusttarget.go#L171).  CRDs, Generators, and Vars also come after.  I guess at some point we should create some tests that demonstrate this as well.

Reply:
Yes, transformers should just work because they are only accumulated after resources (and now components) - see [this part of the code](https://github.com/kubernetes-sigs/kustomize/blob/c5c53011da134543c604d0d8047d22cce1a0aa12/api/internal/target/kusttarget.go#L171).  CRDs, Generators, and Vars also come after.  I guess at some point we should create some tests that demonstrate this as well.